File: apps/sandbox-runtime/internal/tunnel/token_exchange.go
package tunnel

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"
)

const (
	tunnelTokenExchangeRouteSuffix   = "/token-exchange"
	tunnelTokenExchangeRetryDelayMin = time.Second
	tunnelTokenExchangeRetryDelayMax = 30 * time.Second
)

type tunnelTokens struct {
	mutex               sync.RWMutex
	bootstrapToken      string
	tunnelExchangeToken string
}

type tunnelTokenExchangeLoopInput struct {
	Context      context.Context
	GatewayWSURL string
	HTTPClient   *http.Client
	Tokens       *tunnelTokens
}

type tunnelTokenExchangeResponse struct {
	BootstrapToken      string `json:"bootstrapToken"`
	TunnelExchangeToken string `json:"tunnelExchangeToken"`
}

type tunnelTokenExchangeError struct {
	err       error
	retryable bool
}

func (err tunnelTokenExchangeError) Error() string {
	return err.err.Error()
}

func (err tunnelTokenExchangeError) Unwrap() error {
	return err.err
}

func newTunnelTokens(bootstrapToken string, tunnelExchangeToken string) (*tunnelTokens, error) {
	normalizedBootstrapToken, err := normalizeTunnelTokenValue("bootstrap", bootstrapToken)
	if err != nil {
		return nil, err
	}
	normalizedTunnelExchangeToken, err := normalizeTunnelTokenValue("exchange", tunnelExchangeToken)
	if err != nil {
		return nil, err
	}

	return &tunnelTokens{
		bootstrapToken:      normalizedBootstrapToken,
		tunnelExchangeToken: normalizedTunnelExchangeToken,
	}, nil
}

func normalizeTunnelTokenValue(tokenKind string, token string) (string, error) {
	normalizedToken := strings.TrimSpace(token)
	if normalizedToken == "" {
		return "", fmt.Errorf("sandbox tunnel %s token is required", tokenKind)
	}

	return normalizedToken, nil
}

func (tokens *tunnelTokens) CurrentBootstrapToken() string {
	tokens.mutex.RLock()
	defer tokens.mutex.RUnlock()

	return tokens.bootstrapToken
}

func (tokens *tunnelTokens) CurrentTunnelExchangeToken() string {
	tokens.mutex.RLock()
	defer tokens.mutex.RUnlock()

	return tokens.tunnelExchangeToken
}

func (tokens *tunnelTokens) Replace(bootstrapToken string, tunnelExchangeToken string) error {
	normalizedBootstrapToken, err := normalizeTunnelTokenValue("bootstrap", bootstrapToken)
	if err != nil {
		return err
	}
	normalizedTunnelExchangeToken, err := normalizeTunnelTokenValue("exchange", tunnelExchangeToken)
	if err != nil {
		return err
	}

	tokens.mutex.Lock()
	defer tokens.mutex.Unlock()

	tokens.bootstrapToken = normalizedBootstrapToken
	tokens.tunnelExchangeToken = normalizedTunnelExchangeToken
	return nil
}

func buildTunnelTokenExchangeURL(gatewayWSURL string) (string, error) {
	parsedGatewayURL, err := parseGatewayURL(gatewayWSURL)
	if err != nil {
		return "", err
	}

	switch parsedGatewayURL.Scheme {
	case "ws":
		parsedGatewayURL.Scheme = "http"
	case "wss":
		parsedGatewayURL.Scheme = "https"
	default:
		return "", fmt.Errorf("sandbox tunnel gateway ws url must use ws or wss scheme")
	}

	parsedGatewayURL.Path = strings.TrimRight(parsedGatewayURL.Path, "/") + tunnelTokenExchangeRouteSuffix
	parsedGatewayURL.RawQuery = ""
	parsedGatewayURL.Fragment = ""

	return parsedGatewayURL.String(), nil
}

func parseTunnelTokenJWTWindow(token string) (time.Time, time.Time, error) {
	normalizedToken, err := normalizeTunnelTokenValue("exchange", token)
	if err != nil {
		return time.Time{}, time.Time{}, err
	}

	tokenSegments := strings.Split(normalizedToken, ".")
	if len(tokenSegments) != 3 {
		return time.Time{}, time.Time{}, fmt.Errorf("sandbox tunnel exchange token must be a JWT")
	}

	payloadBytes, err := base64.RawURLEncoding.DecodeString(tokenSegments[1])
	if err != nil {
		return time.Time{}, time.Time{}, fmt.Errorf("failed to decode sandbox tunnel exchange token payload: %w", err)
	}

	var payload struct {
		Exp *int64 `json:"exp"`
		Iat *int64 `json:"iat"`
	}
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return time.Time{}, time.Time{}, fmt.Errorf("failed to parse sandbox tunnel exchange token payload: %w", err)
	}
	if payload.Iat == nil {
		return time.Time{}, time.Time{}, fmt.Errorf("sandbox tunnel exchange token iat claim is required")
	}
	if payload.Exp == nil {
		return time.Time{}, time.Time{}, fmt.Errorf("sandbox tunnel exchange token exp claim is required")
	}

	issuedAt := time.Unix(*payload.Iat, 0).UTC()
	expiresAt := time.Unix(*payload.Exp, 0).UTC()
	if !expiresAt.After(issuedAt) {
		return time.Time{}, time.Time{}, fmt.Errorf("sandbox tunnel exchange token exp claim must be after iat")
	}

	return issuedAt, expiresAt, nil
}

func nextTunnelTokenExchangeDelay(now time.Time, issuedAt time.Time, expiresAt time.Time) (time.Duration, error) {
	if !expiresAt.After(issuedAt) {
		return 0, fmt.Errorf("sandbox tunnel exchange token exp claim must be after iat")
	}

	renewAt := issuedAt.Add(expiresAt.Sub(issuedAt) * 4 / 5)
	if !renewAt.After(now) {
		return 0, nil
	}

	return renewAt.Sub(now), nil
}

func nextTunnelTokenExchangeRetryDelay(attempt int) time.Duration {
	delay := tunnelTokenExchangeRetryDelayMin
	for retryIndex := 1; retryIndex < attempt; retryIndex++ {
		if delay >= tunnelTokenExchangeRetryDelayMax/2 {
			return tunnelTokenExchangeRetryDelayMax
		}
		delay *= 2
	}

	return delay
}

func isRetryableTunnelTokenExchangeStatus(statusCode int) bool {
	return statusCode == http.StatusRequestTimeout ||
		statusCode == http.StatusTooManyRequests ||
		statusCode >= http.StatusInternalServerError
}

func shouldRetryTunnelTokenExchange(err error) bool {
	var exchangeErr tunnelTokenExchangeError
	return errors.As(err, &exchangeErr) && exchangeErr.retryable
}

func exchangeTunnelTokens(ctx context.Context, httpClient *http.Client, exchangeURL string, tunnelExchangeToken string) (tunnelTokenExchangeResponse, error) {
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, exchangeURL, bytes.NewReader(nil))
	if err != nil {
		return tunnelTokenExchangeResponse{}, fmt.Errorf("failed to create sandbox tunnel token exchange request: %w", err)
	}
	request.Header.Set("authorization", "Bearer "+tunnelExchangeToken)

	response, err := httpClient.Do(request)
	if err != nil {
		return tunnelTokenExchangeResponse{}, tunnelTokenExchangeError{
			err:       fmt.Errorf("sandbox tunnel token exchange request failed: %w", err),
			retryable: true,
		}
	}
	defer response.Body.Close()

	if response.StatusCode != http.StatusOK {
		return tunnelTokenExchangeResponse{}, tunnelTokenExchangeError{
			err:       fmt.Errorf("sandbox tunnel token exchange request failed with status %d", response.StatusCode),
			retryable: isRetryableTunnelTokenExchangeStatus(response.StatusCode),
		}
	}

	var payload tunnelTokenExchangeResponse
	decoder := json.NewDecoder(response.Body)
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&payload); err != nil {
		return tunnelTokenExchangeResponse{}, fmt.Errorf("failed to decode sandbox tunnel token exchange response: %w", err)
	}

	if err := payload.validate(); err != nil {
		return tunnelTokenExchangeResponse{}, err
	}

	return payload, nil
}

func (response tunnelTokenExchangeResponse) validate() error {
	if _, err := normalizeTunnelTokenValue("bootstrap", response.BootstrapToken); err != nil {
		return fmt.Errorf("sandbox tunnel token exchange response bootstrap token is invalid: %w", err)
	}
	if _, err := normalizeTunnelTokenValue("exchange", response.TunnelExchangeToken); err != nil {
		return fmt.Errorf("sandbox tunnel token exchange response exchange token is invalid: %w", err)
	}

	return nil
}

func runTunnelTokenExchangeLoop(input tunnelTokenExchangeLoopInput) error {
	if input.Context == nil {
		return fmt.Errorf("sandbox tunnel token exchange context is required")
	}
	if input.HTTPClient == nil {
		return fmt.Errorf("sandbox tunnel token exchange http client is required")
	}
	if input.Tokens == nil {
		return fmt.Errorf("sandbox tunnel token exchange tokens are required")
	}

	exchangeURL, err := buildTunnelTokenExchangeURL(input.GatewayWSURL)
	if err != nil {
		return err
	}

	for {
		currentExchangeToken := input.Tokens.CurrentTunnelExchangeToken()
		issuedAt, expiresAt, err := parseTunnelTokenJWTWindow(currentExchangeToken)
		if err != nil {
			return err
		}

		delay, err := nextTunnelTokenExchangeDelay(time.Now().UTC(), issuedAt, expiresAt)
		if err != nil {
			return err
		}

		if delay > 0 {
			timer := time.NewTimer(delay)
			select {
			case <-input.Context.Done():
				if !timer.Stop() {
					<-timer.C
				}
				return nil
			case <-timer.C:
			}
		} else {
			select {
			case <-input.Context.Done():
				return nil
			default:
			}
		}

		for retryAttempt := 1; ; retryAttempt++ {
			if input.Context.Err() != nil {
				return nil
			}
			if !time.Now().UTC().Before(expiresAt) {
				return fmt.Errorf("sandbox tunnel exchange token expired before renewal succeeded")
			}

			exchangeResponse, err := exchangeTunnelTokens(
				input.Context,
				input.HTTPClient,
				exchangeURL,
				currentExchangeToken,
			)
			if err == nil {
				if _, _, parseErr := parseTunnelTokenJWTWindow(exchangeResponse.TunnelExchangeToken); parseErr != nil {
					return parseErr
				}
				if replaceErr := input.Tokens.Replace(
					exchangeResponse.BootstrapToken,
					exchangeResponse.TunnelExchangeToken,
				); replaceErr != nil {
					return replaceErr
				}
				break
			}

			if input.Context.Err() != nil {
				return nil
			}
			if !shouldRetryTunnelTokenExchange(err) {
				return err
			}

			retryDelay := nextTunnelTokenExchangeRetryDelay(retryAttempt)
			remainingUntilExpiry := time.Until(expiresAt)
			if remainingUntilExpiry <= 0 {
				return fmt.Errorf("sandbox tunnel exchange token expired after retryable exchange failures: %w", err)
			}
			if retryDelay > remainingUntilExpiry {
				retryDelay = remainingUntilExpiry
			}

			timer := time.NewTimer(retryDelay)
			select {
			case <-input.Context.Done():
				if !timer.Stop() {
					<-timer.C
				}
				return nil
			case <-timer.C:
			}
		}
	}
}

File: apps/sandbox-runtime/src/tunnel/token-exchange.ts
import { systemScheduler } from "@mistle/time";

import { logSandboxRuntimeEvent } from "../runtime/logger.js";

const TUNNEL_TOKEN_EXCHANGE_ROUTE_SUFFIX = "/token-exchange";
const TUNNEL_RECONNECT_DELAY_MS = 1_000;

type TunnelTokenExchangeResponse = {
  bootstrapToken: string;
  tunnelExchangeToken: string;
};

export class TunnelTokenExchangeError extends Error {
  constructor(message: string) {
    super(message);
  }
}

export class TunnelTokens {
  #bootstrapToken: string;
  #tunnelExchangeToken: string;

  constructor(bootstrapToken: string, tunnelExchangeToken: string) {
    this.#bootstrapToken = normalizeTunnelTokenValue("bootstrap", bootstrapToken);
    this.#tunnelExchangeToken = normalizeTunnelTokenValue("exchange", tunnelExchangeToken);
  }

  currentBootstrapToken(): string {
    return this.#bootstrapToken;
  }

  currentTunnelExchangeToken(): string {
    return this.#tunnelExchangeToken;
  }

  replace(bootstrapToken: string, tunnelExchangeToken: string): void {
    this.#bootstrapToken = normalizeTunnelTokenValue("bootstrap", bootstrapToken);
    this.#tunnelExchangeToken = normalizeTunnelTokenValue("exchange", tunnelExchangeToken);
  }
}

function normalizeTunnelTokenValue(tokenKind: string, token: string): string {
  const normalizedToken = token.trim();
  if (normalizedToken.length === 0) {
    throw new Error(`sandbox tunnel ${tokenKind} token is required`);
  }

  return normalizedToken;
}

export function buildTunnelTokenExchangeUrl(gatewayWsUrl: string): string {
  const parsedGatewayUrl = parseGatewayUrl(gatewayWsUrl);
  parsedGatewayUrl.protocol = parsedGatewayUrl.protocol === "wss:" ? "https:" : "http:";
  parsedGatewayUrl.pathname =
    parsedGatewayUrl.pathname.replace(/\/+$/, "") + TUNNEL_TOKEN_EXCHANGE_ROUTE_SUFFIX;
  parsedGatewayUrl.search = "";
  parsedGatewayUrl.hash = "";

  return parsedGatewayUrl.toString();
}

export function parseGatewayUrl(gatewayWsUrl: string): URL {
  let parsedUrl: URL;
  try {
    parsedUrl = new URL(gatewayWsUrl);
  } catch (error) {
    throw new Error(
      `failed to parse sandbox tunnel gateway ws url: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  if (parsedUrl.protocol !== "ws:" && parsedUrl.protocol !== "wss:") {
    throw new Error("sandbox tunnel gateway ws url must use ws or wss scheme");
  }

  return parsedUrl;
}

export function normalizeBootstrapToken(bootstrapToken: string): string {
  const normalizedToken = bootstrapToken.trim();
  if (normalizedToken.length === 0) {
    throw new Error("sandbox tunnel bootstrap token is required");
  }

  return normalizedToken;
}

export function nextTunnelReconnectDelay(): number {
  return TUNNEL_RECONNECT_DELAY_MS;
}

export function parseTunnelTokenJwtWindow(token: string): {
  issuedAt: Date;
  expiresAt: Date;
} {
  const normalizedToken = normalizeTunnelTokenValue("exchange", token);
  const tokenSegments = normalizedToken.split(".");
  if (tokenSegments.length !== 3) {
    throw new Error("sandbox tunnel exchange token must be a JWT");
  }

  const payloadSegment = tokenSegments[1];
  if (payloadSegment === undefined) {
    throw new Error("sandbox tunnel exchange token must be a JWT");
  }

  let payloadBytes: Buffer;
  try {
    payloadBytes = Buffer.from(payloadSegment, "base64url");
  } catch (error) {
    throw new Error(
      `failed to decode sandbox tunnel exchange token payload: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  let payload: unknown;
  try {
    payload = JSON.parse(payloadBytes.toString("utf8"));
  } catch (error) {
    throw new Error(
      `failed to parse sandbox tunnel exchange token payload: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  const issuedAtSeconds =
    typeof payload === "object" && payload !== null && "iat" in payload ? payload.iat : undefined;
  const expiresAtSeconds =
    typeof payload === "object" && payload !== null && "exp" in payload ? payload.exp : undefined;
  if (typeof issuedAtSeconds !== "number") {
    throw new Error("sandbox tunnel exchange token iat claim is required");
  }
  if (typeof expiresAtSeconds !== "number") {
    throw new Error("sandbox tunnel exchange token exp claim is required");
  }

  const issuedAt = new Date(issuedAtSeconds * 1_000);
  const expiresAt = new Date(expiresAtSeconds * 1_000);
  if (!(expiresAt > issuedAt)) {
    throw new Error("sandbox tunnel exchange token exp claim must be after iat");
  }

  return { issuedAt, expiresAt };
}

export function nextTunnelTokenExchangeDelay(now: Date, issuedAt: Date, expiresAt: Date): number {
  if (!(expiresAt > issuedAt)) {
    throw new Error("sandbox tunnel exchange token exp claim must be after iat");
  }

  const renewAt = new Date(
    issuedAt.getTime() + ((expiresAt.getTime() - issuedAt.getTime()) * 4) / 5,
  );
  if (!(renewAt > now)) {
    return 0;
  }

  return renewAt.getTime() - now.getTime();
}

function validateTunnelTokenExchangeResponse(response: TunnelTokenExchangeResponse): void {
  normalizeTunnelTokenValue("bootstrap", response.bootstrapToken);
  normalizeTunnelTokenValue("exchange", response.tunnelExchangeToken);
}

export async function exchangeTunnelTokens(
  gatewayWsUrl: string,
  tunnelExchangeToken: string,
): Promise<TunnelTokenExchangeResponse> {
  const response = await fetch(buildTunnelTokenExchangeUrl(gatewayWsUrl), {
    method: "POST",
    headers: {
      authorization: `Bearer ${tunnelExchangeToken}`,
    },
  }).catch((error: unknown) => {
    throw new TunnelTokenExchangeError(
      `sandbox tunnel token exchange request failed: ${error instanceof Error ? error.message : String(error)}`,
    );
  });

  if (response.status !== 200) {
    throw new TunnelTokenExchangeError(
      `sandbox tunnel token exchange request failed with status ${String(response.status)}`,
    );
  }

  let payload: unknown;
  try {
    payload = await response.json();
  } catch (error) {
    throw new Error(
      `failed to decode sandbox tunnel token exchange response: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
    throw new Error("failed to decode sandbox tunnel token exchange response: expected object");
  }

  const bootstrapToken = Object.getOwnPropertyDescriptor(payload, "bootstrapToken")?.value;
  const rotatedExchangeToken = Object.getOwnPropertyDescriptor(
    payload,
    "tunnelExchangeToken",
  )?.value;
  if (typeof bootstrapToken !== "string" || typeof rotatedExchangeToken !== "string") {
    throw new Error("sandbox tunnel token exchange response is invalid");
  }

  const normalizedResponse = {
    bootstrapToken,
    tunnelExchangeToken: rotatedExchangeToken,
  };
  validateTunnelTokenExchangeResponse(normalizedResponse);

  return normalizedResponse;
}

export async function exchangeTunnelTokensNow(
  gatewayWsUrl: string,
  tokens: TunnelTokens,
): Promise<void> {
  const currentExchangeToken = tokens.currentTunnelExchangeToken();
  const exchangeResponse = await exchangeTunnelTokens(gatewayWsUrl, currentExchangeToken);
  parseTunnelTokenJwtWindow(exchangeResponse.tunnelExchangeToken);
  tokens.replace(exchangeResponse.bootstrapToken, exchangeResponse.tunnelExchangeToken);
}

export async function runTunnelTokenExchangeLoop(input: {
  signal: AbortSignal;
  gatewayWsUrl: string;
  tokens: TunnelTokens;
  now?: () => Date;
  sleep?: (delayMs: number, signal: AbortSignal) => Promise<void>;
}): Promise<void> {
  const now = input.now ?? (() => new Date());
  const sleep =
    input.sleep ??
    ((delayMs: number, signal: AbortSignal) =>
      new Promise<void>((resolve, reject) => {
        const timeout = systemScheduler.schedule(() => {
          signal.removeEventListener("abort", abortListener);
          resolve();
        }, delayMs);
        const abortListener = (): void => {
          systemScheduler.cancel(timeout);
          reject(signal.reason ?? new Error("operation was aborted"));
        };

        if (signal.aborted) {
          abortListener();
          return;
        }

        signal.addEventListener("abort", abortListener, { once: true });
      }));

  while (!input.signal.aborted) {
    const { issuedAt, expiresAt } = parseTunnelTokenJwtWindow(
      input.tokens.currentTunnelExchangeToken(),
    );
    const exchangeDelay = nextTunnelTokenExchangeDelay(now(), issuedAt, expiresAt);
    await sleep(exchangeDelay, input.signal);

    try {
      await exchangeTunnelTokensNow(input.gatewayWsUrl, input.tokens);
    } catch (error) {
      logSandboxRuntimeEvent({
        level: "warn",
        event: "sandbox_tunnel_token_exchange_failed",
        fields: {
          retryDelayMs: nextTunnelReconnectDelay(),
          message:
            error instanceof Error
              ? error.message
              : typeof error === "string"
                ? error
                : "unknown token exchange error",
        },
      });
      await sleep(nextTunnelReconnectDelay(), input.signal);
    }
  }
}
